roxctl/utils/formatting_writer.go

```diff
@@ -71,46 +57,56 @@ func (w *formattingWriter) SetIndent(indent ...int) {
 
 // WriteString writes the string to the underlying writer, with the configured
 // indentation and wrapping.
-// Implements the StringWriter interface.
-func (w *formattingWriter) WriteString(s string) (int, error) {
-	w.written = 0
-	tokenScanner := bufio.NewScanner(strings.NewReader(s))
-	tokenScanner.Split(wordsAndDelimeters)
-	var err error
-	for err == nil && tokenScanner.Scan() {
-		token := tokenScanner.Text()
-		length := len(token)
-		switch token {
-		case "\t":
-			length = defaultTabWidth
-		case "\n":
-			if w.currentLine == 0 {
-				w.indent.popNotLast()
-				w.indentReset = false
-			}
-			err = w.ln()
-			continue
+func (w *formattingWriter) WriteString(s string) error {
+	for token := range w.tokens(strings.NewReader(s)) {
+		if _, err := w.raw.WriteString(token); err != nil {
+			return err //nolint:wrapcheck
 		}
-		ln := false
-		if w.currentLine == 0 || w.indentReset {
-			if ln, err = w.writePadding(); err != nil {
-				break
-			}
-		}
-		// Write a new line and a padding in case of line overflow:
-		if ln || w.currentLine+length > w.width {
-			if err = w.ln(); err != nil {
-				break
-			}
-			if token == " " {
-				// Do not write space token that caused line break.
+	}
+	return nil
+}
+
+// tokens is an internal iterator that issues tokens with paddings and wrapping.
+func (w *formattingWriter) tokens(r io.Reader) iter.Seq[string] {
+	return func(yield func(string) bool) {
+		tokenScanner := bufio.NewScanner(r)
+		tokenScanner.Split(wordsAndDelimeters)
+		for tokenScanner.Scan() {
+			token := tokenScanner.Text()
+			tokenLength := len(token)
+			switch token {
+			case "\t":
+				tokenLength = w.tabWidth
+			case "\n":
+				if w.currentLine == 0 {
+					w.indent.popNotLast()
+					w.indentReset = false
+				}
+				w.currentLine = 0
+				if !yield("\n") {
+					return
+				}
 				continue
 			}
-			if _, err = w.writePadding(); err != nil {
-				break
+			ln, padding := w.computePadding()
+			// Wrapping condition:
+			if ln || w.currentLine+padding+tokenLength > w.width {
+				w.currentLine = 0
+				if !yield("\n") {
+					return
+				}
+				if token == " " {
+					// Ignore the space that caused wrapping.
+					continue
+				}
+				// Re-compute padding after wrapping.
+				_, padding = w.computePadding()
+			}
+			w.currentLine += padding + tokenLength
+			if (padding > 0 && !yield(strings.Repeat(" ", padding))) ||
+				!yield(token) {
```

Comment:
and remove the condition?
```suggestion
				!yield(strings.Repeat(" ", padding) + token) {
```

Reply:
I tried to avoid unnecessary string concatenations with the `token`, so that the token is yielded as is.
This change would also require unit tests changes, as now more data is yielded before an error is returned from the caller.